VideoLocker/src/main/java/org/edx/mobile/http/ErrorHandlingCallback.java

```diff
@@ -0,0 +1,266 @@
+     */
+    @NonNull
+    private final Type type;
+
+    /**
+     * The callback to invoke on start and finish of the request.
+     */
+    @Nullable
+    private final TaskProgressCallback progressCallback;
+
+    /**
+     * The callback to invoke for delivering any error messages.
+     */
+    @Nullable
+    private final TaskMessageCallback messageCallback;
+
+    /**
+     * Create a new instance of this class.
+     *
+     * @param context A Context for resolving the error message strings. Note that for convenience,
+     *                this will be checked to determine whether it's implementing any of the
+     *                callback interfaces, and will be registered as such if so. If this is not the
+     *                desired outcome, then one of the alternative constructors should be used
+     *                instead, with the relevant callback parameters explicitly passed as null (this
+     *                may require casting the null in case of ambiguity when using a constructor
+     *                that only sets one callback explicitly).
+     * @param type The request type for resolving the type of error messages to deliver.
+     */
+    public ErrorHandlingCallback(@NonNull final Context context,
+                                 @NonNull final Type type) {
+        this(context, type,
+                context instanceof TaskProgressCallback ? (TaskProgressCallback) context : null,
+                context instanceof TaskMessageCallback ? (TaskMessageCallback) context : null);
+    }
+
+    /**
+     * Create a new instance of this class.
+     *
+     * @param context A Context for resolving the error message strings. Note that for convenience,
+     *                this will be checked to determine whether it's implementing the
+     *                {@link TaskMessageCallback} interface, and will be registered as such if so.
+     *                If this is not the desired outcome, then the other constructor should be used
+     *                that takes this callback parameter, and it should be explicitly set as null.
+     * @param type The request type for resolving the type of error messages to deliver.
+     * @param progressCallback The callback to invoke on start and finish of the request. Note that
+     *                         since no callback method in this class is invoked upon request
+     *                         initiation, it assumes that it's being initiated immediately, and
+     *                         thus invokes that start callback immediately as well.
+     */
+    public ErrorHandlingCallback(@NonNull final Context context,
+                                 @NonNull final Type type,
+                                 @Nullable final TaskProgressCallback progressCallback) {
+        this(context, type,
+                progressCallback,
+                context instanceof TaskMessageCallback ? (TaskMessageCallback) context : null);
```

Comment:
No, the `Context` doesn't have to implement any of the listener interfaces, nor does there necessarily need to be listeners registered for every query. This automatic listener type checking and registration is just a convenience functionality, and not really the best practice, since this is a non-obvious side-effect, and it also defeats strong typing. I just carried it over from the previous implementation because updating it to a more robust model would have required careful changes to the implementations everywhere, and we should be moving to the MVP model anyway, which should obviate these semi-hardcoded callback handling.